peer/peer_test.go

```diff
@@ -0,0 +1,145 @@
+)
+
+// A struct that implements AuthInfo interface and implements CommonAuthInfo() method.
+type testAuthInfo struct {
+	credentials.CommonAuthInfo
+}
+
+func (ta testAuthInfo) AuthType() string {
+	return "testAuthInfo"
+}
+
+func TestPeerSecurityLevel(t *testing.T) {
+	testCases := []struct {
+		authLevel credentials.SecurityLevel
+		testLevel credentials.SecurityLevel
+		want      bool
+	}{
+		{
+			authLevel: credentials.PrivacyAndIntegrity,
+			testLevel: credentials.PrivacyAndIntegrity,
+			want:      true,
+		},
+		{
+			authLevel: credentials.IntegrityOnly,
+			testLevel: credentials.PrivacyAndIntegrity,
+			want:      false,
+		},
+		{
+			authLevel: credentials.IntegrityOnly,
+			testLevel: credentials.NoSecurity,
+			want:      true,
+		},
+		{
+			authLevel: credentials.InvalidSecurityLevel,
+			testLevel: credentials.IntegrityOnly,
+			want:      true,
+		},
+		{
+			authLevel: credentials.InvalidSecurityLevel,
+			testLevel: credentials.PrivacyAndIntegrity,
+			want:      true,
+		},
+	}
+	for _, tc := range testCases {
+		ctx := NewContext(context.Background(), &Peer{AuthInfo: testAuthInfo{credentials.CommonAuthInfo{SecurityLevel: tc.authLevel}}})
+		p, ok := FromContext(ctx)
+		if !ok {
+			t.Fatalf("Unable to get peer from context")
+		}
+		err := credentials.CheckSecurityLevel(p.AuthInfo, tc.testLevel)
+		if tc.want && (err != nil) {
+			t.Fatalf("CheckSeurityLevel(%s, %s) returned failure but want success", tc.authLevel.String(), tc.testLevel.String())
+		} else if !tc.want && (err == nil) {
+			t.Fatalf("CheckSeurityLevel(%s, %s) returned success but want failure", tc.authLevel.String(), tc.testLevel.String())
+		}
+	}
+}
+
+type addr struct {
+	ipAddress string
+}
+
+func (addr) Network() string   { return "" }
+func (a *addr) String() string { return a.ipAddress }
+
+func TestPeerStringer(t *testing.T) {
+	testCases := []struct {
+		addr      *addr
+		authLevel credentials.SecurityLevel
+		want      string
+	}{
+		{
+			addr:      &addr{"example.com:1234"},
+			authLevel: credentials.PrivacyAndIntegrity,
+			want:      "Peer{Addr: 'example.com:1234', AuthInfo: 'testAuthInfo'}",
+		},
+		{
+			addr:      &addr{"1.2.3.4:1234"},
+			authLevel: -1,
+			want:      "Peer{Addr: '1.2.3.4:1234', AuthInfo: <nil>}",
+		},
+		{
+			authLevel: credentials.InvalidSecurityLevel,
+			want:      "Peer{Addr: <nil>, AuthInfo: 'testAuthInfo'}",
+		},
+		{
+			authLevel: -1,
+			want:      "Peer{Addr: <nil>, AuthInfo: <nil>}",
+		},
+	}
+	for _, tc := range testCases {
+		ctx := NewContext(context.Background(), &Peer{Addr: tc.addr, AuthInfo: testAuthInfo{credentials.CommonAuthInfo{SecurityLevel: tc.authLevel}}})
+		p, ok := FromContext(ctx)
+		if tc.authLevel == -1 {
+			p.AuthInfo = nil
+		}
+		if tc.addr == nil {
+			p.Addr = nil
+		}
+		if !ok {
+			t.Fatalf("Unable to get peer from context")
+		}
+		if p.String() != tc.want {
+			t.Fatalf("Error using peer String(): expected %q, got %q", tc.want, p.String())
+		}
+	}
```

Comment:
I'm not sure if i understand this pattern here. Why dont we just create a Peer object and test p.String to tc.want

and testCases should just be 

```go
testCases := []struct {
	p	peer 
	want	string
}{
```